extensions/cap_oper.c

```diff
@@ -0,0 +1,152 @@
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
+ *  USA
+ */
+
+#include <stdinc.h>
+#include <modules.h>
+#include <capability.h>
+#include <s_conf.h>
+#include <s_serv.h>
+#include <s_newconf.h>
+#include <client.h>
+#include <msgbuf.h>
+
+static char cap_oper_desc[] = "Provides the solanum.chat/oper capability";
+
+static bool cap_oper_oper_visible(struct Client *);
+static void cap_oper_outbound_msgbuf(void *);
+static void cap_oper_umode_changed(void *);
+static void cap_oper_cap_change(void *);
+
+static unsigned CLICAP_OPER;
+static unsigned CLICAP_OPER_AUSPEX;
+static unsigned CLICAP_OPER_JUSTOPER;
+static unsigned CLICAP_OPER_NORMAL;
+
+static struct ClientCapability capdata_oper_oper = {
+	.visible = cap_oper_oper_visible,
+};
+
+mapi_cap_list_av2 cap_oper_caps[] = {
+	{ MAPI_CAP_CLIENT, "solanum.chat/oper", NULL, &CLICAP_OPER },
+	{ MAPI_CAP_CLIENT, "?oper_auspex", &capdata_oper_oper, &CLICAP_OPER_AUSPEX },
+	{ MAPI_CAP_CLIENT, "?oper_justoper", &capdata_oper_oper, &CLICAP_OPER_JUSTOPER },
+	{ MAPI_CAP_CLIENT, "?oper_normal", &capdata_oper_oper, &CLICAP_OPER_NORMAL },
+	{ 0, NULL, NULL, NULL },
+};
+
+mapi_hfn_list_av1 cap_oper_hfnlist[] = {
+	{ "outbound_msgbuf", cap_oper_outbound_msgbuf, HOOK_NORMAL },
+	{ "umode_changed", cap_oper_umode_changed, HOOK_MONITOR },
+	{ "cap_change", cap_oper_cap_change, HOOK_MONITOR },
+	{ NULL, NULL, 0 },
+};
+
+static bool
+cap_oper_oper_visible(struct Client *client)
+{
+	return false;
+}
+
+static void
+cap_oper_outbound_msgbuf(void *data_)
+{
+	hook_data *data = data_;
+	struct MsgBuf *msgbuf = data->arg1;
+
+	if (data->client == NULL || !IsPerson(data->client))
+		return;
+
+	if (IsOper(data->client))
+	{
+		/* send all oper data to auspex */
+		msgbuf_append_tag(msgbuf, "solanum.chat/oper", data->client->user->opername, CLICAP_OPER_AUSPEX);
+		if (HasPrivilege(data->client, "oper:hidden"))
+			/* these people aren't allowed to see hidden opers */
+			return;
+		msgbuf_append_tag(msgbuf, "solanum.chat/oper", data->client->user->opername, CLICAP_OPER_JUSTOPER);
+		msgbuf_append_tag(msgbuf, "solanum.chat/oper", NULL, CLICAP_OPER_NORMAL);
+	}
+}
+
+static inline void
+update_clicap_oper(struct Client *client)
+{
+	/* clear out old caps */
+	client->localClient->caps &= ~CLICAP_OPER_AUSPEX;
+	client->localClient->caps &= ~CLICAP_OPER_JUSTOPER;
+	client->localClient->caps &= ~CLICAP_OPER_NORMAL;
+
+	if (client->localClient->caps & CLICAP_OPER && HasPrivilege(client, "auspex:oper"))
+	{
+		/* if the client is an oper with auspex, let them see everything */
+		client->localClient->caps |= CLICAP_OPER_AUSPEX;
+	}
+	else if (ConfigFileEntry.hide_opers)
```

Comment:
I think this needs to be in `cap_oper_outbound_msgbuf`, since it's a configuration setting that could change at runtime.